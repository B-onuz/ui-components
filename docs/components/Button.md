# Button (Component)

Componente de botão, pode ser usados em formulários e para outros tipos de ação.

## Importação

```js
import Button from 'ui-components/lib/components/Button'
```

## Como usar

```jsx
<Button>Enviar</Button>
```

## Variações

### Colors

```jsx
<Button color={'primary'}>Enviar</Button>
```

Escala no tema: `theme.colors`

### Size

```jsx
<Button size={'small'}>Enviar</Button>
```

Valores: `small` `medium` `big` `huge`

### Outras props

```jsx
<Button color={'primary'} type={'submit'}>
  Salvar
</Button>
```

Este componente recebe outras propriedades da tag button, como a propriedade `type`.

## Styled-system props

- space https://styled-system.com/table#space
