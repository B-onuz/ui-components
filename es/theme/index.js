"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var colors = {
  fullDarkRed: "#550e33",
  darkRed: "#9c2543",
  lightRed: "#ef4d5a",
  red: "#db2828",
  orange: "#f2711c",
  yellow: "#fbbd08",
  olive: "#b5cc18",
  green: "#21BA45",
  teal: "#00b5ad",
  blue: "#2185d0",
  violet: "#6435c9",
  purple: "#a333c8",
  pink: "#e03997",
  brown: "#a5673f",
  grey: "#767676",
  black: "#1b1c1d"
};
colors.primary = colors.lightRed;
colors.secondary = colors.fullDarkRed;
colors["default"] = '#e0e1e2';
var fontSizes = [12, 14, 16, 20, 24, 32];
var _default = {
  fontFamily: "'Lato', sans-serif",
  colors: colors,
  fontSizes: fontSizes
};
exports["default"] = _default;