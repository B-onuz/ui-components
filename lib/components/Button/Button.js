"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _color = _interopRequireDefault(require("color"));

var _styledSystem = require("styled-system");

var _variants = require("./variants");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _propTypes2 = _interopRequireDefault(require("@styled-system/prop-types"));

var _Spinner = _interopRequireDefault(require("../Spinner"));

var _this = void 0,
    _jsxFileName = "/Users/rogeralbino/projects/Kazap/B-onuz/ui-components/src/components/Button/Button.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  appearance: none;\n  font-family: inherit;\n  border-radius: 4px;\n  outline: none;\n  cursor: pointer;\n  padding: .5rem 1rem;\n  border-style: solid;\n  ", "\n  ", "\n  ", "\n  ", "\n  box-sizing: border-box;\n  &:focus {\n    box-shadow: 0 0 8px ", ";\n  }\n  \n  &:active:not(:disabled) {\n    box-shadow: inset 0px 0px 5px #c1c1c1;\n  }\n  &:active,\n  &:hover {\n    background-color: ", ";\n  }\n  &:disabled {\n    background-color: ", ";\n  }\n  ", "\n  \n  ", "\n\n  ", " \n\n  ", " \n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var StyledButton = _styledComponents["default"].button(_templateObject(), _variants.color, _variants.size, _styledSystem.space, function (props) {
  return props.fullWidth && 'width: 100%;' || '';
}, function (_ref) {
  var theme = _ref.theme,
      color = _ref.color;
  return theme.colors[color];
}, function (_ref2) {
  var theme = _ref2.theme,
      color = _ref2.color;
  return new _color["default"](theme.colors[color]).saturate(0.1).darken(0.2).string();
}, function (_ref3) {
  var theme = _ref3.theme,
      color = _ref3.color;
  return new _color["default"](theme.colors[color]).desaturate(0.4).lighten(0.1).opaquer(0.7).string();
}, function (_ref4) {
  var loading = _ref4.loading;
  return !!loading && "\n    cursor: progress;\n  ";
}, function (_ref5) {
  var outline = _ref5.outline,
      theme = _ref5.theme,
      color = _ref5.color;
  return !!outline && "\n    background-color: #fff;\n    color: " + theme.colors[color] + " !important;\n    border-color: " + theme.colors[color] + ";\n    &:hover {\n      background-color: " + theme.colors[color] + ";\n      color: " + (new _color["default"](theme.colors[color]).isDark() ? '#fff' : '#000') + " !important;\n    }\n    ";
}, function (_ref6) {
  var underline = _ref6.underline;
  return !!underline ? "\n      text-decoration: underline;\n    " : "text-decoration: none;";
}, function (_ref7) {
  var linkButton = _ref7.linkButton,
      theme = _ref7.theme,
      color = _ref7.color;
  return !!linkButton ? "\n    color: " + theme.colors[color] + ";\n    border: none;\n    background-color: transparent !important;\n    box-shadow: none !important;\n    outline: none;\n    &:hover {\n      text-decoration: underline;\n      background-color: transparent;\n    }\n    " : "\n     text-decoration: none;\n    ";
});

var Button = function Button(_ref8) {
  var children = _ref8.children,
      loading = _ref8.loading,
      disabled = _ref8.disabled,
      rest = _objectWithoutPropertiesLoose(_ref8, ["children", "loading", "disabled"]);

  return /*#__PURE__*/_react["default"].createElement(StyledButton, _extends({}, rest, {
    disabled: !!disabled || !!loading,
    loading: loading,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 81,
      columnNumber: 3
    }
  }), !!loading ? 'carregando...' : children);
};

Button.displayName = 'Button';
Button.propTypes = process.env.NODE_ENV !== "production" ? _extends({
  size: _propTypes["default"].oneOf(['small', 'medium', 'big', 'huge']),
  color: _propTypes["default"].string
}, _propTypes2["default"].space) : {};
Button.defaultProps = {
  size: 'medium',
  color: 'default',
  type: 'button'
};
var _default = Button;
exports["default"] = _default;
module.exports = exports.default;