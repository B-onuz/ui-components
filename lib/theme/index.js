"use strict";

exports.__esModule = true;
exports["default"] = void 0;
var colors = {
  fullDarkRed: '#550e33',
  darkRed: '#9c2543',
  lightRed: '#ef4d5a',
  mediumRed: '#FF4F5D',
  red: '#db2828',
  orange: '#f2711c',
  yellow: '#fbbd08',
  olive: '#b5cc18',
  green: '#21BA45',
  teal: '#00b5ad',
  blue: '#2185d0',
  violet: '#6435c9',
  purple: '#a333c8',
  pink: '#e03997',
  brown: '#a5673f',
  grey: '#767676',
  darkGrey: '#9A9A9A',
  black: '#1b1c1d',
  warning: '#7A4D05',
  success: '#1E561F',
  info: '#0E566C',
  danger: '#973937',
  cyan: '#1CB5AD'
};
colors.primary = colors.lightRed;
colors.secondary = colors.fullDarkRed;
colors["default"] = '#e0e1e2';
var breakpoints = ['40em', '52em', '64em', '80em'];
breakpoints.sm = breakpoints[0];
breakpoints.md = breakpoints[1];
breakpoints.lg = breakpoints[2];
breakpoints.xl = breakpoints[3];
var fontSizes = [12, 14, 16, 20, 24, 32];
var _default = {
  fontFamily: "'Lato', sans-serif",
  colors: colors,
  fontSizes: fontSizes,
  breakpoints: breakpoints
};
exports["default"] = _default;
module.exports = exports.default;