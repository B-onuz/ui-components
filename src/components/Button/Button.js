import React from "react";
import styled from "styled-components";
import Color from "color";
import { color, variant, space } from "styled-system";

const Button = styled.button`
  appearance: none;
  font-family: inherit;
  border-radius: 4px;
  outline: none;
  cursor: pointer;
  padding: .5rem 1rem;
  ${props => variant({
    variants: {
      primary: {
        color: new Color(props.theme.colors.lightRed).isDark() ? '#fff' : '#000',
        bg: "lightRed",
        fontSize: 2,
      },
      secondary: {
        color: new Color(props.theme.colors.fullDarkRed).isDark() ? '#fff' : '#000',
        bg: "fullDarkRed",
        fontSize: 2,
      },
      default: {
        color: new Color(props.theme.colors[props.bg] || '#e0e1e2').isDark() ? '#fff' : '#000',
        bg: "#e0e1e2",
        fontSize: 2,
      },
    },
  })}
  ${color}
  ${space}
  border: 1px solid ${({ theme, variant, bg }) => bg ? theme.colors[bg] : theme.colors[variant] };
  &:focus {
    box-shadow: 0 0 8px ${({ theme, variant, bg }) => bg ? theme.colors[bg] : theme.colors[variant] || "rgba(0,0,0,.3)"};
  }
  &:hover {
    background-color: ${({ theme, variant, bg }) => bg ? new Color(theme.colors[bg]).lighten(0.2).string() : new Color(theme.colors[variant] || "#ddd").lighten(0.15).string()};
  }
  &:active {
    background-color: ${({ theme, variant, bg }) => bg ? new Color(theme.colors[bg]).darken(0.2).string() : new Color(theme.colors[variant] || "#ddd").darken(0.15).string()};
  }
`;

export default Button;
