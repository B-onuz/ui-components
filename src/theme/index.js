export default {
  colors: {
    fullDarkRed: "#550e33",
    darkRed: "#9c2543",
    lightRed: "#ef4d5a",
    red: "#db2828",
    orange: "#f2711c",
    yellow: "fbbd08",
    olive: "#b5cc18"
  }
}
