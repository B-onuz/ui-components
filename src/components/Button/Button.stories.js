import React from 'react'
import Button from './Button'

export default {
  title: 'Button',
  component: Button
};

export const ButtonVariants = () => (
  <div>
    <Button m={2} variant={'default'}>{'Default'}</Button>
    <Button m={2} variant={'primary'}>{'Primary'}</Button>
    <Button m={2} variant={'secondary'}>{'Secondary'}</Button>
  </div>
);

export const ButtonColors = () => (
  <div>
    <Button m={2} variant={'default'} bg={'fullDarkRed'} >{'fullDarkRed'}</Button>
    <Button m={2} variant={'default'} bg={'darkRed'} >{'darkRed'}</Button>
    <Button m={2} variant={'default'} bg={'lightRed'} >{'lightRed'}</Button>
    <Button m={2} variant={'default'} bg={'red'} >{'red'}</Button>
    <Button m={2} variant={'default'} bg={'orange'} >{'orange'}</Button>
    <Button m={2} variant={'default'} bg={'yellow'} >{'yellow'}</Button>
    <Button m={2} variant={'default'} bg={'olive'} >{'olive'}</Button>
    <Button m={2} variant={'default'} bg={'green'} >{'green'}</Button>
    <Button m={2} variant={'default'} bg={'teal'} >{'teal'}</Button>
    <Button m={2} variant={'default'} bg={'blue'} >{'blue'}</Button>
    <Button m={2} variant={'default'} bg={'violet'} >{'violet'}</Button>
    <Button m={2} variant={'default'} bg={'purple'} >{'purple'}</Button>
    <Button m={2} variant={'default'} bg={'pink'} >{'pink'}</Button>
    <Button m={2} variant={'default'} bg={'brown'} >{'brown'}</Button>
    <Button m={2} variant={'default'} bg={'grey'} >{'grey'}</Button>
    <Button m={2} variant={'default'} bg={'black'} >{'black'}</Button>
  </div>
);
