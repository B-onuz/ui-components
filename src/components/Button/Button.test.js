import React from 'react'
import Button from './Button.js'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import theme from '../../theme'

test('Should render button', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}><Button>Hello Button</Button></ThemeProvider>
  );
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
});

test('Should render button when bg prop', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Button m={2} variant={'default'} bg={'fullDarkRed'} >{'fullDarkRed'}</Button>
      <Button m={2} variant={'default'} bg={'darkRed'} >{'darkRed'}</Button>
      <Button m={2} variant={'default'} bg={'lightRed'} >{'lightRed'}</Button>
      <Button m={2} variant={'default'} bg={'red'} >{'red'}</Button>
      <Button m={2} variant={'default'} bg={'orange'} >{'orange'}</Button>
      <Button m={2} variant={'default'} bg={'yellow'} >{'yellow'}</Button>
      <Button m={2} variant={'default'} bg={'olive'} >{'olive'}</Button>
      <Button m={2} variant={'default'} bg={'green'} >{'green'}</Button>
      <Button m={2} variant={'default'} bg={'teal'} >{'teal'}</Button>
      <Button m={2} variant={'default'} bg={'blue'} >{'blue'}</Button>
      <Button m={2} variant={'default'} bg={'violet'} >{'violet'}</Button>
      <Button m={2} variant={'default'} bg={'purple'} >{'purple'}</Button>
      <Button m={2} variant={'default'} bg={'pink'} >{'pink'}</Button>
      <Button m={2} variant={'default'} bg={'brown'} >{'brown'}</Button>
      <Button m={2} variant={'default'} bg={'grey'} >{'grey'}</Button>
      <Button m={2} variant={'default'} bg={'black'} >{'black'}</Button>
    </ThemeProvider>
  );
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
});

test('Should render button when variant prop', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Button m={2} variant={'default'}>{'Default'}</Button>
      <Button m={2} variant={'primary'}>{'Primary'}</Button>
      <Button m={2} variant={'secondary'}>{'Secondary'}</Button>
    </ThemeProvider>
  );
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
});