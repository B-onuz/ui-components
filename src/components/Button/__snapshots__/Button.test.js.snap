// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render button 1`] = `
<button
  className="sc-AxjAm eNOgXJ"
>
  Hello Button
</button>
`;

exports[`Should render button when bg prop 1`] = `
Array [
  <button
    className="sc-AxjAm jTOLzc"
  >
    fullDarkRed
  </button>,
  <button
    className="sc-AxjAm lloiYN"
  >
    darkRed
  </button>,
  <button
    className="sc-AxjAm hWJeKR"
  >
    lightRed
  </button>,
  <button
    className="sc-AxjAm cNZtdV"
  >
    red
  </button>,
  <button
    className="sc-AxjAm eKiVSe"
  >
    orange
  </button>,
  <button
    className="sc-AxjAm eKtUwP"
  >
    yellow
  </button>,
  <button
    className="sc-AxjAm QmVOg"
  >
    olive
  </button>,
  <button
    className="sc-AxjAm DNCTV"
  >
    green
  </button>,
  <button
    className="sc-AxjAm febjO"
  >
    teal
  </button>,
  <button
    className="sc-AxjAm gorsKs"
  >
    blue
  </button>,
  <button
    className="sc-AxjAm gxDmOi"
  >
    violet
  </button>,
  <button
    className="sc-AxjAm iPKdnM"
  >
    purple
  </button>,
  <button
    className="sc-AxjAm grgtEz"
  >
    pink
  </button>,
  <button
    className="sc-AxjAm hkXJio"
  >
    brown
  </button>,
  <button
    className="sc-AxjAm gXaUQg"
  >
    grey
  </button>,
  <button
    className="sc-AxjAm enSPxy"
  >
    black
  </button>,
]
`;

exports[`Should render button when variant prop 1`] = `
Array [
  <button
    className="sc-AxjAm eernPf"
  >
    Default
  </button>,
  <button
    className="sc-AxjAm heuNxC"
  >
    Primary
  </button>,
  <button
    className="sc-AxjAm YtSgw"
  >
    Secondary
  </button>,
]
`;
